emscripten/native/graphics/BitmapCanvasEmscriptenImpl.cpp

```diff
@@ -0,0 +1,63 @@
+#include "graphics/BitmapCanvasEmscriptenImpl.h"
+#include "components/Exceptions.h"
+#include "utils/BitmapUtils.h"
+
+#include <cmath>
+
+namespace carto {
+    BitmapCanvas::EmscriptenImpl::EmscriptenImpl(int width, int height)
+    {
+        
```

Comment:
Canvas is used internally to render `Text` and `Popup` vector elements. Any chance implementing this using HMTL5 canvas/embind?

Reply:
@mtehver, I implemented it but it was very difficult. Popup is running on main thread but text is not. Firstly I wanted to use Canvas but I couldn't access any dom element on background thread. After that I tried to use OffscreenCanvas. I implemented it but app froze. I think main thread was blocked.
Related issue: https://github.com/emscripten-core/emscripten/issues/17130

To fix this issue I moved main loop to background thread using OffscreenCanvas. It worked. Main problem was that Safari didn't support OffscreenCanvas. At the same time, the listeners didn't work on background thread. So I examined emscripten proxy and thread source code to find out how it works on main thread. I found it but related methods are in internal header and BitmapCanvas::EmscriptenImpl is protected. So i had to write a bit of a complex code to use these functions.

As a result, the application can be used on both main thread and background thread. It works in Safari, Chrome and Firefox. It works on desktop and mobile.